tool/teleport-update/main.go

```diff
@@ -0,0 +1,351 @@
+func writeUpdatesConfig(filename string, cfg *UpdatesConfig) error {
+	opts := append([]renameio.Option{
+		renameio.WithPermissions(0755),
+		renameio.WithExistingPermissions(),
+	})
+	t, err := renameio.NewPendingFile(filename, opts...)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+	defer t.Cleanup()
+	err = yaml.NewEncoder(t).Encode(cfg)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+	return trace.Wrap(t.CloseAtomicallyReplace())
+}
+
+// cmdDisable disables updates.
+func cmdDisable(ctx context.Context, ccfg *cliConfig) error {
+	var (
+		versionsDir = filepath.Join(ccfg.DataDir, versionsDirName)
+		updatesYAML = filepath.Join(versionsDir, updatesFileName)
+	)
+	unlock, err := lock(filepath.Join(versionsDir, lockFileName), false)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+	defer unlock()
+	cfg, err := readUpdatesConfig(updatesYAML)
+	if err != nil {
+		return trace.Errorf("failed to read updates.yaml: %w", err)
+	}
+	if !cfg.Spec.Enabled {
+		plog.InfoContext(ctx, "Automatic updates already disabled")
+		return nil
+	}
+	cfg.Spec.Enabled = false
+	if err := writeUpdatesConfig(updatesYAML, cfg); err != nil {
+		return trace.Errorf("failed to write updates.yaml: %w", err)
+	}
+	return nil
+}
+
+// cmdEnable enables updates and triggers an initial update.
+func cmdEnable(ctx context.Context, ccfg *cliConfig) error {
+	return trace.NotImplemented("TODO")
+}
+
+// cmdUpdate updates Teleport to the version specified by cluster reachable at the proxy address.
+func cmdUpdate(ctx context.Context, ccfg *cliConfig) error {
+	return trace.NotImplemented("TODO")
+}
+
+func lock(lockFile string, nonblock bool) (func(), error) {
+	if err := os.MkdirAll(filepath.Dir(lockFile), 0755); err != nil {
+		return nil, trace.Wrap(err)
+	}
+	lf, err := os.OpenFile(lockFile, os.O_CREATE|os.O_WRONLY|os.O_APPEND, 0666)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	how := syscall.LOCK_EX
+	if nonblock {
+		how |= syscall.LOCK_NB
+	}
+	if err := syscall.Flock(int(lf.Fd()), how); err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	return func() {
+		if err := syscall.Flock(int(lf.Fd()), syscall.LOCK_UN); err != nil {
+			plog.Debug("Failed to unlock lock file", "file", lockFile, "error", err)
+		}
+		if err := lf.Close(); err != nil {
+			plog.Debug("Failed to close lock file", "file", lockFile, "error", err)
+		}
+	}, nil
+}
+
+type downloadConfig struct {
+	// Insecure turns off TLS certificate verification when enabled.
+	Insecure bool
+	// Pool defines the set of root CAs to use when verifying server
+	// certificates.
+	Pool *x509.CertPool
+	// Timeout is a timeout for requests.
+	Timeout time.Duration
+}
+
+func newClient(cfg *downloadConfig) (*http.Client, error) {
+	rt := apiutils.NewHTTPRoundTripper(&http.Transport{
+		TLSClientConfig: &tls.Config{
+			InsecureSkipVerify: cfg.Insecure,
+			RootCAs:            cfg.Pool,
+		},
+		Proxy: func(req *http.Request) (*url.URL, error) {
+			return httpproxy.FromEnvironment().ProxyFunc()(req.URL)
+		},
```

Comment:
Do we gain much from using x/net/http/httpproxy over just `http.ProxyFromEnvironment`?

Reply:
I copied the webclient for consistency: https://github.com/gravitational/teleport/blob/f2b0d11c3c2bd3fc5131e940c9f199c647342e78/api/client/webclient/webclient.go#L92-L112

Reply:
We should definitely update the webclient too, that's quite inefficient - the standard library uses a `sync.Once` to store the returned `ProxyFunc()` from the vendored copy of `x/net/http/httpproxy`.

FWIW I would just take the `lib/defaults.Transport()` and add the `TLSClientConfig` to it unless there's a reason to avoid that.

Reply:
The webclient uses `tracehttp` -- should I avoid that?

Reply:
Switched to the defaults transport to merge -- let me know if I should bring tracehttp back in the future. Not sure If we need the telemetry provided by that package.